filebeat/fileset/pipelines_test.go

```diff
@@ -0,0 +1,101 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// +build !integration
+
+package fileset
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/elastic/beats/libbeat/outputs/elasticsearch"
+)
+
+func TestLoadPipelinesWithMultiPipelineFileset(t *testing.T) {
+	tests := []struct {
+		name        string
+		esVersion   string
+		expectedErr string
+	}{
+		{
+			name:        "ES < 6.5.0",
+			esVersion:   "6.4.1",
+			expectedErr: "the mod/fls fileset has multiple pipelines, which are only supported with Elasticsearch >= 6.5.0. Currently running with Elasticsearch version 6.4.1",
+		},
+		{
+			name:        "ES == 6.5.0",
+			esVersion:   "6.5.0",
+			expectedErr: "",
+		},
+		{
+			name:        "ES > 6.5.0",
+			esVersion:   "6.6.0",
+			expectedErr: "",
+		},
+	}
+
+	for _, test := range tests {
```

Comment:
This loop creates a many concurrent HTTP servers not being cleaned up between the different tests. This is due to `defer` operating at function-scope, not block scope.

Use `t.Run(name, func(t *testing.T) { ... })` so to create separate isolated tests (defer in the body will properly cleanup resources). This improves output/readability of test results, plus allows developers to run single tests.

Common pattern:

```
cases := map[string]struct{
   ...
}{
  ...
}
for name, test := range cases {
  test := test // copy test into local block scope (in case you want to enable t.Parallel())
  t.Run(name, func(t *testing.T) {
    // common test body
  })
}
```

Reply:
@urso Would you recommend using `t.Parallel()` in general, or would you say it depends on how much work each test case is doing and how many test cases there are to run in all?

Reply:
Running with `t.Parallel()` is a great way to see if things are really isolated :)
Tests are run package by package. Using `t.Parallel()` is a nice way to speed up tests in a package that already a few seconds. Checking libbeat unit tests, most are finished in a few ms.

When using `t.Parallel()` the parent test blocks until all it's child tests have returned. The number of active concurrent tests to be run depends on the `-parallel` CLI flag. by default it's the number of cores (GOMAXPROCS env variable) you have in your system. We don't use t.Parallel() much, but for packages with a large number of unit tests, that also take a little longer to execute it can really make a difference. The `t.Run()` method always spawns a go-routine. That is, there is no real additional cost by adding `t.Parallel()` even for very small tests. 

Just tried to enable t.Parallel() on beats queues test on my machine. Duration went from 20s to ~5s. I'm always happy if I don't have to wait for long unit tests to finish.

We should also make more use of `testing/quick` (if applicable). These will definitely profit from `t.Parallel()`.

Still most time is spend in system tests.